File: 2020/kickstart/c/perfect-subarray.py
import math
import bisect

def read_ints():
    return list(map(int, input().split(' ')))

MAXT = 100 * 100000
m = [0] * 2 * MAXT

def solve():
    T = int(input())
    for t in range(1, T + 1):
        n = int(input())
        nums = read_ints()

        mi, ma = 0, -(1 << 32)
        s = 0
        for x in nums:
            s += x
            mi = min(x, mi, s)
            ma = max(x, ma, s)
        
        R = int(math.sqrt(ma - mi))
        ks = [i * i for i in range(R + 1)]

        table = m.copy()
        count = 0
        s = 0
        mins = 0
        table[MAXT] = 1
        for x in nums:
            s += x
            mins = min(mins, s)
            kn = min(bisect.bisect_right(ks, s - mins) + 1, len(ks))
            for j in range(kn):
                k = ks[j]
                count += table[MAXT + s - k]
            table[s + MAXT] += 1
        print('Case #{}: {}'.format(t, count))
        
solve()

'''
1
4
-10 -2 11 5

4
3
2 2 6
5
30 30 9 1 30
4
4 0 0 16
4
-10 -2 11 5
'''

File: 2020/kickstart/c/perfect-subarray.cpp
#include <iostream>
#include <unordered_map>
#include <algorithm>
#include <cmath>
#include <cstring>

using namespace std;

const int MAXN = 100005;
const int MAXT = 100 * MAXN;
int nums[MAXN];
int table[2 * MAXT];

int main()
{
    int T;
    scanf("%d", &T);

    for (int t = 1; t <= T; t += 1)
    {
        int n;
        scanf("%d", &n);

        int s = 0, mi = 0, ma = 0;

        for (int i = 0; i < n; i += 1)
        {
            scanf("%d", nums + i);
            if (nums[i] >= 0) ma += nums[i];
            else mi -= nums[i];
        }

        int R = (int)sqrt(mi + ma);
        int* ks = new int[R + 1];
        for (int i = 0; i <= R; i += 1)
        {
            ks[i] = i * i;
        }

        memset(table, 0, 2 * MAXT * sizeof(int));

        table[MAXT] = 1;
        s = 0;
        long long count = 0;
        for (int i = 0; i < n; i += 1)
        {
            s += nums[i];
            for (int j = 0; j <= R; j += 1)
            {
                count += table[s - ks[j] + MAXT];
            }
            table[s + MAXT] += 1;
        }

        cout << "Case #" << t << ": " << count << endl;

        delete ks;
    }
}

/*
3
3
2 2 6
5
30 30 9 1 30
4
4 0 0 16

1
3 
2 2 6
*/